jsonschema/validators.py

```diff
@@ -862,6 +868,44 @@ def resolve_remote(self, uri):
             self.store[uri] = result
         return result
 
+    def store_subschema(self, schema, last_schema=None, last_url=None):
+        """
+        Using $id or id with $ref, save subschema to self.store
+
+        Arguments:
+
+            schema:
+
+                The referring schema.
+
+            last_schema:
+
+                The referring upper level schema.
+
+            last_url:
+
+                Save the last URL.
+        """
+        if not isinstance(schema, dict) \
+                or self.resolution_scope in list_schema:
+            return
+
+        for k in schema.keys():
+            if k in [u"id", u"$id"] and isinstance(schema[k], str_types):
+
+                last_url = urljoin(last_url, schema[k], allow_fragments=True)
+                url, fragment = urldefrag(last_url)
+
+                if last_schema:
+                    self.store[url] = last_schema
+
+                if fragment:
+                    self.store[url][fragment] = schema
```

Comment:
Add fragment element and save it into `url": { last_schema }` which saved above, in the form of  `"url": { last_schema, "fagment": { schema } }`